src/main/java/gregtech/common/covers/CoverItemFilter.java

```diff
@@ -32,25 +31,69 @@
 import codechicken.lib.render.pipeline.IVertexOperation;
 import codechicken.lib.vec.Cuboid6;
 import codechicken.lib.vec.Matrix4;
+import com.cleanroommc.modularui.drawable.Rectangle;
+import com.cleanroommc.modularui.factory.SidedPosGuiData;
+import com.cleanroommc.modularui.screen.ModularPanel;
+import com.cleanroommc.modularui.utils.Alignment;
+import com.cleanroommc.modularui.value.sync.EnumSyncValue;
+import com.cleanroommc.modularui.value.sync.GuiSyncManager;
+import com.cleanroommc.modularui.widgets.SlotGroupWidget;
+import com.cleanroommc.modularui.widgets.layout.Column;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
 
 public class CoverItemFilter extends CoverBase implements CoverWithUI {
 
     protected final String titleLocale;
     protected final SimpleOverlayRenderer texture;
-    protected final ItemFilterWrapper itemFilter;
+    protected final ItemFilterContainer itemFilterContainer;
     protected ItemFilterMode filterMode = ItemFilterMode.FILTER_INSERT;
     protected ItemHandlerFiltered itemHandler;
 
     public CoverItemFilter(@NotNull CoverDefinition definition, @NotNull CoverableView coverableView,
-                           @NotNull EnumFacing attachedSide, String titleLocale, SimpleOverlayRenderer texture,
-                           ItemFilter itemFilter) {
+                           @NotNull EnumFacing attachedSide, String titleLocale, SimpleOverlayRenderer texture) {
         super(definition, coverableView, attachedSide);
         this.titleLocale = titleLocale;
         this.texture = texture;
-        this.itemFilter = new ItemFilterWrapper(this);
-        this.itemFilter.setItemFilter(itemFilter);
-        this.itemFilter.setMaxStackSize(1);
+        this.itemFilterContainer = new ItemFilterContainer(this);
+    }
+
+    @Override
+    public void onAttachment(@NotNull CoverableView coverableView, @NotNull EnumFacing side,
+                             @Nullable EntityPlayer player, @NotNull ItemStack itemStack) {
+        super.onAttachment(coverableView, side, player, itemStack);
+        var dropStack = GTUtility.copy(1, itemStack);
+        this.itemFilterContainer.setFilterStack(dropStack);
+    }
+
+    @Override
+    public @NotNull ItemStack getPickItem() {
+        if (TooltipHelper.isCtrlDown())
+            return getCoverableView().getStackForm();
+
+        return this.itemFilterContainer.getFilterStack();
+    }
+
+    @Override
+    public void writeInitialSyncData(@NotNull PacketBuffer packetBuffer) {
+        packetBuffer.writeByte(this.filterMode.ordinal());
+        packetBuffer.writeBoolean(itemFilterContainer.hasFilter());
```

Comment:
Shouldn't this write if there is a filter first, so that you can save the read of the filterMode until after checking if a filter exists?